common/src/app/common/files/migrations.cljc

```diff
@@ -1145,6 +1146,36 @@
         (update :pages-index update-vals update-container)
         (update :components update-vals update-container))))
 
+(defn migrate-up-60
+  [data]
+  (letfn [(update-component
+            [component]
+            (let [objects   (:objects component) ;; we only have encounter this on deleted components,
+                                                ;; so the relevant objects are inside the component
+                  cycles-ids (->> objects
+                                  vals
+                                  (filter #(= (:id %) (:shape-ref %)))
+                                  (map :id))
+                  to-detach (->> cycles-ids
+                                 (map #(get objects %))
+                                 (map #(ctn/get-head-shape objects %))
+                                 (map :id)
+                                 distinct
+                                 (mapcat #(ctn/get-children-in-instance objects %))
+                                 (map :id)
+                                 set)]
+              (update component :objects
+                      (fn [objects]
+                        (reduce-kv (fn [acc k v]
+                                     (if (contains? to-detach k)
+                                       (assoc acc k (ctk/detach-shape v))
+                                       (assoc acc k v)))
+                                   {}
+                                   objects)))))]
+
+    (-> data
+        (update :components update-vals update-component))))
```

Comment:
I'd add a filter to process only deleted components. For more speed, and also to avoid the possibility of the code to have unintended effects in a component without `:objects`.